x/virtualgroup/keeper/msg_server.go

```diff
@@ -0,0 +1,390 @@
+	}
+
+	depositDenom := k.DepositDenomForGVG(ctx)
+	if req.Withdraw.Denom != depositDenom {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidDenom, "invalid coin denomination: got %s, expected %s", req.Withdraw.Denom, k.DepositDenomForGVG(ctx))
+	}
+
+	var withdrawTokens math.Int
+
+	availableTokens := k.GetAvailableStakingTokens(ctx, gvg)
+	if req.Withdraw.Amount.IsZero() {
+		withdrawTokens = availableTokens
+	} else {
+		if availableTokens.LT(req.Withdraw.Amount) {
+			return nil, types.ErrWithdrawAmountTooLarge
+		}
+		withdrawTokens = req.Withdraw.Amount
+	}
+
+	// withdraw the deposit token from module account to funding account.
+	coins := sdk.NewCoins(sdk.NewCoin(depositDenom, withdrawTokens))
+	if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, sp.GetFundingAccAddress(), coins); err != nil {
+		return nil, err
+	}
+
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventUpdateGlobalVirtualGroup{
+		Id:           req.GlobalVirtualGroupId,
+		StoreSize:    gvg.StoredSize,
+		TotalDeposit: gvg.TotalDeposit,
+	}); err != nil {
+		return nil, err
+	}
+
+	return &types.MsgWithdrawResponse{}, nil
+}
+
+func (k msgServer) SwapOut(goCtx context.Context, req *types.MsgSwapOut) (*types.MsgSwapOutResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	operatorAddr := sdk.MustAccAddressFromHex(req.OperatorAddress)
+	sp, found := k.spKeeper.GetStorageProviderByOperatorAddr(ctx, operatorAddr)
+	if !found {
+		return nil, sptypes.ErrStorageProviderNotFound
+	}
+
+	successorSP, found := k.spKeeper.GetStorageProvider(ctx, req.SuccessorSpId)
+	if !found {
+		return nil, sptypes.ErrStorageProviderNotFound.Wrapf("successor sp not found.")
+	}
+
+	// verify the approval
+	err := gnfdtypes.VerifySignature(sdk.MustAccAddressFromHex(successorSP.ApprovalAddress), sdk.Keccak256(req.GetApprovalBytes()), req.SuccessorSpApproval.Sig)
+	if err != nil {
+		return nil, err
+	}
+
+	if req.GlobalVirtualGroupFamilyId == types.NoSpecifiedFamilyId {
+		// if the family id is not specified, it means that the SP will swap out as a secondary SP.
+		err := k.SwapOutAsSecondarySP(ctx, sp.Id, successorSP.Id, req.GlobalVirtualGroupIds)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// if the family id is specified, it means that the SP will swap out as a primary SP and the successor sp will
+		// take over all the gvg of this family
+		err := k.SwapOutAsPrimarySP(ctx, sp, successorSP, req.GlobalVirtualGroupFamilyId)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventSwapOut{
+		StorageProviderId:          sp.Id,
+		GlobalVirtualGroupFamilyId: req.GlobalVirtualGroupFamilyId,
+		GlobalVirtualGroupIds:      req.GlobalVirtualGroupIds,
+		SuccessorSpId:              successorSP.Id,
+	}); err != nil {
+		return nil, err
+	}
+	return &types.MsgSwapOutResponse{}, nil
+}
+
+func (k msgServer) Settle(goCtx context.Context, req *types.MsgSettle) (*types.MsgSettleResponse, error) {
```

Comment:
what about just settling in `GlobalVirtualGroupFamilyId` level? Simple and I think most of the case the SP will use this.

Reply:
Seem it will be not very easy for secondary sp - they have to ask Primary SP to get their income.